cis194/week10/zhansongl/AParser.hs

```diff
@@ -0,0 +1,67 @@
+{-# OPTIONS_GHC -Wall #-}
+
+module AParser where
+
+import Data.Char (isDigit, isUpper)
+import Control.Applicative (Alternative(..))
+
+newtype Parser a
+  = Parser { runParser :: String -> Maybe (a, String) }
+
+satisfy :: (Char -> Bool) -> Parser Char
+satisfy p = Parser f
+  where f [] = Nothing    -- fail on the empty input
+        f (x:xs)          -- check if x satisfies the predicate
+                          -- if so, return x along with the remainder
+                          -- of the input (that is, xs)
+            | p x       = Just (x, xs)
+            | otherwise = Nothing  -- otherwise, fail
+
+char :: Char -> Parser Char
+char c = satisfy (== c)
+
+posInt :: Parser Integer
+posInt = Parser f
+  where f xs
+          | null ns = Nothing
+          | otherwise = Just (read ns, rest)
+          where (ns, rest) = span isDigit xs
+
+first :: (a -> b) -> (a, c) -> (b, c)
+first f p = (f . fst $ p, snd p)
+
+parserFmap :: (a -> b) -> Parser a -> String -> Maybe (b, String)
+parserFmap f pa s = first f <$> runParser pa s 
+
+parserReplace :: a -> Parser b -> String -> Maybe (a, String)
+parserReplace a pb s = first (const a) <$> runParser pb s
+
+instance Functor Parser where
+  fmap f = Parser . parserFmap f
+  (<$) a = Parser . parserReplace a
```

Comment:
why not:

instance Functor Parser where
  fmap f (Paser g) = Parser $ fmap (first f) . g

Reply:
hmm, good catch, will fix tomorrow

Reply:
😄 I copied from https://github.com/fp-works/2019-winter-Haskell-school/blob/master/cis194/week5/robert-chang/src/Parser.hs#L23

Reply:
done 😎